api/cds.proto

```diff
@@ -417,6 +417,11 @@ message Cluster {
   // For instance, if the metadata is intended for the Router filter, the filter
   // name should be specified as *envoy.router*.
   Metadata metadata = 25;
+
+  // Cluster can operate both HTTP1.1 and HTTP2 connections and the protocol
+  // is chosen based on the downstream protocol, if any. In case there is no
+  // downstream connection, this option will be ignored.
+  bool auto_http2 = 26;
```

Comment:
Assuming we do end up taking this upstream, I think we should make this field more specific, such as `infer_http2_from_downstream_connection` or something like that. I also wonder if this should go inside the `Http2ProtocolOptions` message.

Reply:
The semantics looks complicated and the variable name is too terse, as @mattklein123 suggested.

Can we make this behavior automatic instead of a specific option? Every boolean option essentially double the testing matrix. It adds more work to monitoring and support as well. W need to justify the value before adding another option.

Reply:
Unfortunately I don't think this can be the default behavior.

Reply:
I thought about this more and I would suggest actually making this an enum, with the enum default being the current behavior (effectively, "select based on codec options presence). The reason for this is eventually we will want to support HTTP/1 -> HTTP/2 upgrade and this could be a future enum value. cc @rshriram